### DISCUSSION TITLE
Ray for Debugging

### DISCUSSION THREAD

**Discussion by gchtr** [timber/timber#2399] [General]
Okay, so let’s try out GitHub discussions…

Spatie just [released Ray](https://twitter.com/freekmurze/status/1347197034256887815), a debugging app for PHP and Laravel applications, which looks really promising.

It looks like this can work for debugging Timber and WordPress just as fine.

I feel like this is a strong contender for the [Debugging Guide](https://timber.github.io/docs/guides/debugging/), because it lies somewhere between using Symfony VarDumper and using Xdebug breakpoints inside an IDE. Maybe it works out of the box, maybe it could be combined with https://github.com/timnashcouk/wp-debug-helpers.

In case anyone already tried it out with Timber, tell us how it went!

🤡 Fun fact: They [almost named it Timber](https://twitter.com/freekmurze/status/1347222330720845825). In that case, we could be debugging Timber with Timber now 😄.

**Comment by palmiak** [comment]
Just testing right now :) I really like how it works. I had no problem setting it up on Windows.

In terms of PHP part it just works without any problem. Which is great.

In twig files it's not so easy - using fn ends with an error, but I think that creating a function for twig and a set o filters (to simulate the chaining) would solve the problem. Will try to do it soon-ish.

**Comment by jarednova** [comment]
Agreed ! This looks really promising, and a nice more user-friendly way to follow issues through. I've used XDebug at various times in the past, but also found that it can be a bear to setup.

**Comment by anugupta** [comment]
Also worth noting that Query Monitor has a logger built into it now and uses actions, so doesn't break production if you forget to remove (which i've never ever ever done, of course!)

`do_action( 'qm/debug', 'This happened!' );` 

You can also pass in arrays and QM displays them nicely.

`do_action( 'qm/debug', $context );`

https://querymonitor.com/blog/2018/07/profiling-and-logging/

**Comment by daviddarke** [comment]
👋 Started using Ray for Laravel and WP+Timber development work. Working well. I had to add a little custom Twig filter:

```php
add_filter('timber/twig', 'rayOutput');

function rayOutput($twig){
    if (function_exists('ray')) {
        $twig->addFilter(new Twig_Filter('ray', function ($vriable) {
            ray($vriable);
        }));
    }

    return $twig;
}
```

Then call in twig using:

```twig
{{ post|ray }}
```
